src/dune_engine/build_system.ml

```diff
@@ -1534,17 +1541,25 @@ end = struct
             Action_exec.exec ~root ~context ~env ~targets:(Some targets)
               ~rule_loc:loc ~build_deps ~execution_parameters action
           in
-          let files_in_directory_targets =
+          let produced_targets =
             match sandbox with
-            | None -> Path.Build.Set.empty
+            | None ->
+              (* Directory targets are not allowed for non-sandboxed actions, so
+                 the call below should not raise. *)
+              Targets.Produced.of_validated_files targets ~on_dir_target:`Raise
             | Some sandbox ->
               (* The stamp file for anonymous actions is always created outside
                  the sandbox, so we can't move it. *)
-              let files = remove_stamp_file targets.files rule_kind in
-              Sandbox.move_targets_to_build_dir sandbox ~loc ~files
-                ~dirs:targets.dirs
+              let should_be_skipped =
+                match rule_kind with
+                | Normal_rule -> fun (_ : Path.Build.t) -> false
+                | Anonymous_action { stamp_file; _ } ->
+                  Path.Build.equal stamp_file
+              in
+              Sandbox.move_targets_to_build_dir sandbox ~loc ~should_be_skipped
+                ~targets
```

Comment:
Note for reviewers: the change above is slightly non-trivial. I changed the code to pass `should_be_skipped` instead of filtering out files because this works better with the upcoming PR on promoting directory targets.